# Analysing Used Car Listings on eBay Kleinanzeigen

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which which year the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which which year the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

We will start by importing the pandas and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('C:/Users/imer/Downloads/autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
# Printing info and .head methods about autos

print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

### First observations from dataset

At first glance, one can immediately realise that the dataset contains 50,000 rows of data - with most columns containing data for all rows.

The columns that do not contain the full 50k rows of data are `fuelType`, `notRepairedDamage`, `model`, `gearbox` and `vehicleType`. A decision will be taken later on what to do with such columns.

Another important observation to note is that all column names are written in CamelCase format, rather than the usual snakecase Python format - this will have to be arranged.

The data set contains 20 columns - most of them are strings. 

Let's start cleaning the dataset.



In [4]:
# Converting column names format to snakecase

print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In the output above, it was shown how the dataframe column names were successfully changed to snakecase format as per Python convention. The first few rows of the dataframe were printed to be seen.

Now we will carry on with more data exploration to look for any other data cleaning tasks to be done. Initially, we will look for the following:-
* Text columns where almost/all values are the same. These can often be removed since they don't have useful information for analysis.

* Data which is stored as text data, but can easily be converted to numerical.

We will use the `.describe()`, `value_counts()` and `.head()` methods to help in this data exploration.

In [6]:
# Look at descriptive stats for all columns using describe() method for ALL columns

autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

One can easily note the following:-

* `price` and `odometer` can easily be converted to numerical type (they include the dollar sign and 'km' respectively)
* `nr_of_pictures` column does not consist of a single value - this needs more investigation.
* `seller` and `offer_type` columns only contain 2 unique values - these can also be removed.



In [7]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
print(autos['price'].head())
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)
autos.rename({'odometer' : 'odometer_km'}, axis=1, inplace=True)
print(autos['odometer_km'].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int32


### Data Cleaning Part 2

From the last screen, we learned that there are a number of text columns where almost all of the values are the same (`seller` and `offer_type`). We also converted the `price` and `odometer` columns to numeric types and renamed odometer to `odometer_km`.

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns.

#### `Price` column

We will investigate the following:

* unique values for the column
* min/max/median/etc using the describe() method
* Look for the lowest and the highest values


In [8]:
# unique values:

print('The nr of unique values for the price column is ', autos['price'].unique().shape, 'values')

The nr of unique values for the price column is  (2357,) values


In [9]:
# Descriptive stats for price column

print(autos['price'].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [10]:
# Look for the frequency of any values, first the lowest, then the highest
print('The 20 lowest prices indicated in this column are:')
print(autos['price'].value_counts().sort_index(ascending=True).head(20))
print('\n')
print('The 20 highest prices indicated in this column are:')
print(autos['price'].value_counts().sort_index(ascending=False).head(20))

The 20 lowest prices indicated in this column are:
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


The 20 highest prices indicated in this column are:
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


#### Lowest prices

From the data above, it can be seen that there are 1,421 listings listed at zero price while some listings increase steadily per 1 dollar. Since it is possible that people bid 1 dollar on ebay, we will keep these values but remove the 'free' listings.

#### Highest prices

Values keep rising sensibly up until 350,000 dollars. However, from then onwards there seem to be some unrealistic listings. 
Therefore, it is decided that we will keep the listings with the price range 1 < $ < 350,000.

The code and the resultant description is shown in the below cell.

In [11]:
# Removing the listings with zero price and above 350k:

autos = autos[autos['price'].between(1,351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

#### `odometer_km` column

We will now follow the same procedure for the Odometer column:

In [12]:
# Look for the frequency of any values, first the lowest, then the highest
print('The mileage ranges for these listings are:')
print('\n')
print(autos['odometer_km'].value_counts().sort_index(ascending=False))

The mileage ranges for these listings are:


150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64


It is clear that users had an option to select from a datafield the amount of miles for each listing (rather than manually inputted by the user). This is because all values are rounded up.

### Data Cleaning Part 3 - Date Columns

There are a number of columns with date information:

* `date_crawled`
* `registration_month`
* `registration_year`
* `ad_created`
* `last_seen`

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings. We'll start by printing the first 5 rows of the date columns whose data is crawled from the website.

In [13]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

For each of these columns, the first 10 characters of each row can be attributed to the date in the format 'yyyy/mm/dd'.
Hence, we will extract these values, and count the values for the data range. Rather than frequency, we will obtain values in relative format, hence we will use the `normalize=True` parameter. We also want null values to be included so `dropna=False` will be used.

In [14]:
#Date Crawled column:

(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=True))

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

It seems as though the data was crawled over the period March-April 2016. The distribution is approximately uniform.

In [15]:
# Last Seen column

(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False))

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-31    0.023783
2016-03-12    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [16]:
# Ad Created data

(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_values(ascending=False))

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
                ...   
2016-01-29    0.000021
2016-02-22    0.000021
2016-01-13    0.000021
2016-02-16    0.000021
2015-06-11    0.000021
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

### Registration year data

Next we will understand the distribution for the registration year column.


In [17]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

One can notice that this column also contains some anomaly data, since we have a min of year 1000 (way before cars were invented) and a max of year 9999 (way in the future).

Let's now deal with this data.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [18]:
# using cars reg date between 1900 and 2016

autos = autos[autos['registration_year'].between(1900,2016)]

# Check the top 10 years in which these were registered:
autos['registration_year'].value_counts().head(10)

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
Name: registration_year, dtype: int64

In [19]:
# And relatively:

autos['registration_year'].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

The results above show that the cars listed were mostly registered between 1999 and 2007.


### Explore Price by Brand

We will now see the following:
    
* most popular brand on ebay
* most expensive or cheapest brands

In [20]:
# Most popular brands on ebay

autos['brand'].value_counts().sort_values(ascending=False)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

In [21]:
# The above in relative terms:

autos['brand'].value_counts(normalize=True).sort_values(ascending=False)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German brands are by far the most popular: VW, BMW, Opel and Mercedes Benz all part of the top 5.

For the purpose of our study, we will consider the 15 most popular brands.

We will do this below:

In [22]:
brands = autos['brand'].value_counts(normalize=True)

popular_brands = brands.head(15).index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen'],
      dtype='object')


In [23]:
# We will now look for brand mean and median prices

brand_mean_price = {}
brand_median_price = {}

for b in popular_brands:
    brand_only = autos[autos['brand'] == b]
    mean_price = brand_only['price'].mean()
    median_price = brand_only['price'].median()
    brand_mean_price[b] = int(mean_price)

# Display dictionary of brand with mean prices
brand_mean_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749,
 'renault': 2474,
 'peugeot': 3094,
 'fiat': 2813,
 'seat': 4397,
 'skoda': 6368,
 'nissan': 4743,
 'mazda': 4112,
 'smart': 3580,
 'citroen': 3779}


Of the top 15 brands, there is a distinct price gap:

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive so it might explain their popularity.
* Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.
* Fiat is the cheapest and they find themselves in the top 10.

### Calculating Mean Mileage for the Top 15 Brands

Using the same method as above, we will now look to compute the mean mileage for the top 15 brands.

We will do this to check for any relation between the mean mileage and the mean price for each of these brands.

We will create a dictionary `brand_mean_mileage` to store the results.

In [24]:
brand_mean_mileage = {}  # dictionary to store brand mean mileage results

for m in popular_brands:
    brand_only = autos[autos['brand'] == m]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[m] = int(mean_mileage)
    
brand_mean_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266,
 'renault': 128071,
 'peugeot': 127153,
 'fiat': 117121,
 'seat': 121131,
 'skoda': 110848,
 'nissan': 118330,
 'mazda': 124464,
 'smart': 99326,
 'citroen': 119694}

In [25]:
bmp = pd.Series(brand_mean_price)
print(bmp.sort_values(ascending=False))

audi             9336
mercedes_benz    8628
bmw              8332
skoda            6368
volkswagen       5402
nissan           4743
seat             4397
mazda            4112
citroen          3779
ford             3749
smart            3580
peugeot          3094
opel             2975
fiat             2813
renault          2474
dtype: int64


In [26]:
mmp = pd.Series(brand_mean_mileage)
print(mmp.sort_values(ascending=False))

bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
renault          128071
peugeot          127153
mazda            124464
ford             124266
seat             121131
citroen          119694
nissan           118330
fiat             117121
skoda            110848
smart             99326
dtype: int64


In [27]:
brand_info = pd.DataFrame(mmp, columns=['mean_mileage'])
brand_info

mean_mileage
volkswagen           128707
bmw                  132572
opel                 129310
mercedes_benz        130788
audi                 129157
ford                 124266
renault              128071
peugeot              127153
fiat                 117121
seat                 121131
skoda                110848
nissan               118330
mazda                124464
smart                 99326
citroen              119694

In [28]:
brand_info['mean_price $'] = bmp
brand_info.sort_values(by=['mean_price $'], ascending=False)

mean_mileage  mean_price $
audi                 129157          9336
mercedes_benz        130788          8628
bmw                  132572          8332
skoda                110848          6368
volkswagen           128707          5402
nissan               118330          4743
seat                 121131          4397
mazda                124464          4112
citroen              119694          3779
ford                 124266          3749
smart                 99326          3580
peugeot              127153          3094
opel                 129310          2975
fiat                 117121          2813
renault              128071          2474

### Conclusion on Mileage vs Price

From the above data, it can be deduced that mileage doesn't have much effect on the price that the cars is listed for. In fact, Audi has one of the highest mileage average from the top 15 brands, despite it being the most expensive car, on average.

In fact, based on mileage and price **ONLY**, buyers might be better off buying an Opel, Renault or Peugeot since they show a much better price with a similar mileage to the top 5 brands.

In [29]:
autos['ad_created'] = autos['ad_created'].replace('-','')
autos['ad_created']

0        2016-03-26 00:00:00
1        2016-04-04 00:00:00
2        2016-03-26 00:00:00
3        2016-03-12 00:00:00
4        2016-04-01 00:00:00
                ...         
49995    2016-03-27 00:00:00
49996    2016-03-28 00:00:00
49997    2016-04-02 00:00:00
49998    2016-03-08 00:00:00
49999    2016-03-13 00:00:00
Name: ad_created, Length: 46681, dtype: object

### Data Cleaning - Next Steps:

In the next steps we are going to:

* translate any German words to English
* convert date format 'yyyy-mm-dd' to numeric

In [30]:
# # Translate german words to english:

print('unique values for fuel type are ', autos['fuel_type'].unique())
print('unique values for vehicle type are ', autos['vehicle_type'].unique())
print('unique values for unrepaired damage are ', autos['unrepaired_damage'].unique())

unique values for fuel type are  ['lpg' 'benzin' 'diesel' nan 'cng' 'hybrid' 'elektro' 'andere']
unique values for vehicle type are  ['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']
unique values for unrepaired damage are  ['nein' nan 'ja']


In [31]:
autos['fuel_type'] = autos['fuel_type'].replace('benzin', 'petrol').replace('elektro', 'electrical').replace('andere','other')
autos['vehicle_type'] = autos['vehicle_type'].replace('kleinwagen', 'small car').replace('kombi', 'estate').replace('andere','other')
autos['unrepaired_damage'] = autos['unrepaired_damage'].replace('nein', 'no').replace('ja','yes')
autos['model'] = autos['model'].replace('andere', 'other')
autos['fuel_type']
autos['vehicle_type']
autos['unrepaired_damage']

0        no
1        no
2        no
3        no
4        no
         ..
49995    no
49996    no
49997    no
49998    no
49999    no
Name: unrepaired_damage, Length: 46681, dtype: object

In [32]:
# Convert date to numeric type for ad_created, last_seen, date_crawled

import datetime as dt
#ad created
autos['ad_created'] = pd.to_datetime(autos['ad_created']).dt.strftime('%Y%m%d')
print(autos['ad_created'])

#date created
autos['date_crawled'] = pd.to_datetime(autos['date_crawled']).dt.strftime('%Y%m%d')
print(autos['date_crawled'])

#last seen
autos['last_seen'] = pd.to_datetime(autos['last_seen']).dt.strftime('%Y%m%d')
print(autos['last_seen'])

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160313
Name: ad_created, Length: 46681, dtype: object
0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160314
Name: date_crawled, Length: 46681, dtype: object
0        20160406
1        20160406
2        20160406
3        20160315
4        20160401
           ...   
49995    20160401
49996    20160402
49997    20160404
49998    20160405
49999    20160406
Name: last_seen, Length: 46681, dtype: object


### Data Analysis part 2 - Comparing brand vs model


In [33]:
brands_vs_models = {} # empty dictionary to store results
autos.dropna(subset=['model']) # drop rows with no value in 'model' column

#unique values for model and brand
unique_models = autos['model'].unique()
unique_brands = autos['brand'].unique()

for b in unique_brands:
    selected_rows = autos[autos['brand'] == b]
    sorted_rows = selected_rows.sort_values('model', ascending=False)
    top_model = sorted_rows.iloc[0]
    top_models = top_model['model']
    brands_vs_models[b] = top_models

print('These are the most popular models by brand on Ebay:')
brands_vs_models

These are the most popular models by brand on Ebay:


{'peugeot': 'other',
 'bmw': 'z_reihe',
 'volkswagen': 'up',
 'smart': 'roadster',
 'ford': 'transit',
 'chrysler': 'voyager',
 'seat': 'toledo',
 'renault': 'twingo',
 'mercedes_benz': 'vito',
 'audi': 'tt',
 'sonstige_autos': nan,
 'opel': 'zafira',
 'mazda': 'rx_reihe',
 'porsche': 'other',
 'mini': 'other',
 'toyota': 'yaris',
 'dacia': 'sandero',
 'nissan': 'x_trail',
 'jeep': 'wrangler',
 'saab': 'other',
 'volvo': 'xc_reihe',
 'mitsubishi': 'pajero',
 'jaguar': 'x_type',
 'fiat': 'stilo',
 'skoda': 'yeti',
 'subaru': 'other',
 'kia': 'sportage',
 'citroen': 'other',
 'chevrolet': 'spark',
 'hyundai': 'tucson',
 'honda': 'other',
 'daewoo': 'other',
 'suzuki': 'swift',
 'trabant': 'other',
 'land_rover': 'range_rover_sport',
 'alfa_romeo': 'spider',
 'lada': 'samara',
 'rover': 'rangerover',
 'daihatsu': 'terios',
 'lancia': 'ypsilon'}

### Data Analysis Part 3 - Mean Price of Non/Damaged Cars

Now we will investigate the price of cars with any damage which has been left unrepaired vs those with no damage.

In [44]:
damaged = autos['price'][autos['unrepaired_damage'] == 'yes'].mean()
print('The mean price of cars with damage that is still to be repaired is $',damaged)
print('\n')

undamaged_repair = autos['price'][autos['unrepaired_damage'] == 'no'].mean()
print('The mean price of cars without damage that is still to be repaired is $',undamaged_repair)

The mean price of cars with damage that is still to be repaired is $ 2241.146035242291


The mean price of cars without damage that is still to be repaired is $ 7164.033102796004
